File: server/src/test/java/com/fluffy/support/testcontainers/PostgreSQLContainerInitializer.java
package com.fluffy.support.testcontainers;

import java.util.Map;
import org.jetbrains.annotations.NotNull;
import org.springframework.boot.test.util.TestPropertyValues;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.ConfigurableApplicationContext;
import org.testcontainers.containers.PostgreSQLContainer;

public class PostgreSQLContainerInitializer implements ApplicationContextInitializer<ConfigurableApplicationContext> {

    private static final PostgreSQLContainer<?> POSTGRESQL = new PostgreSQLContainer<>("postgres:16-alpine");

    static {
        POSTGRESQL.start();
    }

    @Override
    public void initialize(@NotNull ConfigurableApplicationContext context) {
        Map<String, String> properties = Map.of(
                "spring.datasource.url", POSTGRESQL.getJdbcUrl(),
                "spring.datasource.username", POSTGRESQL.getUsername(),
                "spring.datasource.password", POSTGRESQL.getPassword()
        );

        TestPropertyValues.of(properties).applyTo(context);
    }
}

File: server/src/test/kotlin/com/fluffy/support/testcontainers/PostgreSQLContainerInitializer.kt
package com.fluffy.support.testcontainers

import org.springframework.boot.test.util.TestPropertyValues
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.testcontainers.containers.PostgreSQLContainer

class PostgreSQLContainerInitializer : ApplicationContextInitializer<ConfigurableApplicationContext> {

    companion object {
        private val POSTGRESQL = PostgreSQLContainer("postgres:16.3-alpine")

        init {
            POSTGRESQL.start()
        }
    }

    override fun initialize(context: ConfigurableApplicationContext) {
        val properties = mapOf(
            "spring.datasource.url" to POSTGRESQL.jdbcUrl,
            "spring.datasource.username" to POSTGRESQL.username,
            "spring.datasource.password" to POSTGRESQL.password
        )

        TestPropertyValues.of(properties).applyTo(context)
    }
}